### PR TITLE
add webhook

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/16000

Add mutating and validating webhook.

### CODE REVIEW

File: helm/cluster-api-provider-aws/templates/webhook/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  template:
+    metadata:
+      labels:
+        {{- include "labels.common" . | nindent 8 }}
+    spec:
+      containers:
+      - args:
+        - --metrics-addr=127.0.0.1:8080
+        - --webhook-port=9443
+        - --feature-gates=EKS=${EXP_EKS:=false},MachinePool=${EXP_MACHINE_POOL:=false}
```
Comment (anvddriesch) [inline helm/cluster-api-provider-aws/templates/webhook/deployment.yaml:23]:
```suggestion
        - --feature-gates=EKS={{ .Values.featuregates.eks }},MachinePool={{ .Values.featuregates.machinepool }}
```

**Review by MarcelMue — COMMENTED** [review]
Does it make sense to add this here? I am concerned what happens if we run multiple versions of this controller like this

**Comment by anvddriesch** [comment]
Right now things aren't really working because we miss values in the CR. We want to try adding the webhook to solve that issue.

**Comment by MarcelMue** [comment]
Let's maybe chat this afternoon together how to tackle this :)

**Comment by TheoBrigitte** [comment]
We just trying to create a cluster so far. Let's first figure out if we need webhook at all.

File: helm/cluster-api-provider-aws/templates/webhook/certificate.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+---
+apiVersion: cert-manager.io/v1alpha2
```
Comment (njuettner) [inline helm/cluster-api-provider-aws/templates/webhook/certificate.yaml:2]:
```suggestion
apiVersion: cert-manager.io/v1
```

File: helm/cluster-api-provider-aws/templates/webhook/certificate.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+spec:
+  dnsNames:
+  - {{ include "resource.default.name" . }}.{{ include "resource.webhook.namespace" . }}.svc
+  - {{ include "resource.default.name" . }}.{{ include "resource.webhook.namespace" . }}.svc.cluster.local
+  issuerRef:
+    kind: Issuer
+    name: {{ include "resource.default.name" . }}
+  secretName: {{ include "resource.default.name" . }}-cert
+---
+apiVersion: cert-manager.io/v1alpha2
```
Comment (njuettner) [inline helm/cluster-api-provider-aws/templates/webhook/certificate.yaml:18]:
```suggestion
apiVersion: cert-manager.io/v1
```

**Review by njuettner — COMMENTED** [review]
In case we're using our cert-manager I'd suggest to update the api version to `v1`. This has the potential to break in case we stick with `v1alpha2`.

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml [hunk 4]
```diff
@@ -0,0 +1,199 @@
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: capi-webhook-system/capa-serving-cert
```
Comment (MarcelMue) [inline helm/cluster-api-provider-aws/templates/webhook/webhook.yaml:6]: FIXME

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    - UPDATE
+    resources:
+    - awsmanagedmachinepools
+  sideEffects: None
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: capi-webhook-system/capa-serving-cert
```
Comment (MarcelMue) [inline helm/cluster-api-provider-aws/templates/webhook/webhook.yaml:95]: FIXME

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml (hunk 4 above)
Comment (MarcelMue) [inline helm/cluster-api-provider-aws/templates/webhook/webhook.yaml:6]:
```suggestion
    cert-manager.io/inject-ca-from: {{ include "resource.default.namespace" . }}/{{ include "resource.default.name" . }}-certificates
```

File: helm/cluster-api-provider-aws/templates/webhook/certificate.yaml (hunk 2 above)
Comment (TheoBrigitte) [reply helm/cluster-api-provider-aws/templates/webhook/certificate.yaml]: fixed

File: helm/cluster-api-provider-aws/templates/webhook/certificate.yaml (hunk 3 above)
Comment (TheoBrigitte) [reply helm/cluster-api-provider-aws/templates/webhook/certificate.yaml]: fixed

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml (hunk 5 above)
Comment (TheoBrigitte) [reply helm/cluster-api-provider-aws/templates/webhook/webhook.yaml]: fixed

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml (hunk 4 above)
Comment (TheoBrigitte) [reply helm/cluster-api-provider-aws/templates/webhook/webhook.yaml]: fixed

**Review by MarcelMue — APPROVED** [review]
Works for now

File: helm/cluster-api-provider-aws/templates/webhook/webhook.yaml (hunk 4 above)
Comment (TheoBrigitte) [reply helm/cluster-api-provider-aws/templates/webhook/webhook.yaml]: fixed

**Pull request merged**